src/iter_both.rs

```diff
@@ -0,0 +1,133 @@
+use fastq_pair::{delete_empty_fastq, parse_header, parse_read, PartialRead};
+use std::collections::HashMap;
+use std::fs::File;
+use std::io::{BufReader, BufWriter, Result};
+use std::io::Write;
+use std::path::Path;
+
+/// Contains all Read/Write objects
+struct IO {
+    in_read1: BufReader<File>,
+    in_read2: BufReader<File>,
+    out_read1: BufWriter<File>,
+    out_read2: BufWriter<File>,
+    out_single: BufWriter<File>,
+    singleton_path: String,
+}
+
+/// Create all IO objects for reading and writing
+fn create_io(r1_path: &str, r2_path: &str) -> Result<IO> {
+    let parent = Path::new(r1_path).parent().unwrap();
+    let r1_out_path = parent.join("R1_paired.fastq").to_str().unwrap().to_string();
+    let r2_out_path = parent.join("R2_paired.fastq").to_str().unwrap().to_string();
+    let singleton_path = parent.join("Singletons.fastq").to_str().unwrap().to_string();
+    // Readers
+    let r1_handle = File::open(&r1_path)?;
+    let r2_handle = File::open(&r2_path)?;
+    let r1_reader = BufReader::new(r1_handle);
+    let r2_reader = BufReader::new(r2_handle);
+    // Writers
+    let r1_out_handle = File::create(&r1_out_path)?;
+    let r2_out_handle = File::create(&r2_out_path)?;
+    let singleton_handle = File::create(&singleton_path)?;
+    let r1_writer = BufWriter::new(r1_out_handle);
+    let r2_writer = BufWriter::new(r2_out_handle);
+    let singleton_writer = BufWriter::new(singleton_handle);
+    Ok(IO {
+        in_read1: r1_reader,
+        in_read2: r2_reader,
+        out_read1: r1_writer,
+        out_read2: r2_writer,
+        out_single: singleton_writer,
+        singleton_path: singleton_path.to_string(),
+    })
+}
+
+
+/// Writes out paired reads two FASTQ files
+fn write_read(header: &str,
+              w1: &mut BufWriter<File>,
+              w2: &mut BufWriter<File>,
+              map1: &mut HashMap<String, PartialRead>,
+              map2: &mut HashMap<String, PartialRead>) -> Result<()> {
+    let r1 = &map1[header];
+    let r2 = &map2[header];
+    write!(w1, "{}.1\n{}+\n{}", header, r1.seq, r1.qscore)?;
+    write!(w2, "{}.2\n{}+\n{}", header, r2.seq, r2.qscore)?;
+    &map1.remove(header);
+    &map2.remove(header);
+    Ok(())
+}
+
+/// Pair two FASTQ files by iterating over both files simultaneously.
+/// Should be much more memory efficient than "store_read" method if
+/// files are mostly paired
+pub fn pair_fastqs(r1_path: &str, r2_path: &str) -> Result<()> {
+    let mut io = create_io(r1_path, r2_path)?;
+    let mut map1 = HashMap::new();
+    let mut map2 = HashMap::new();
+
+    // TODO: Iterate until both reads are empty instead of just R1 -- maybe just use a loop with a break
+    while let Some(read1) = parse_read(&mut io.in_read1) {
+        let read2 = parse_read(&mut io.in_read2).expect("Failed to read R2");
```

Comment:
Looks like you're aware of this from the TODO, but this will crash if R1 has fewer singletons than R2.

Reply:
Ya, I should just fix this before merging, I was hoping to come up with something better than using `loop` and breaking before I finished but didn't.